core/src/main/java/io/hyperfoil/core/impl/SimulationRunner.java

```diff
@@ -411,6 +416,117 @@ public void addGlobalData(Map<String, GlobalData.Element> globalData) {
+            long count = bean.getCollectionCount();
+            if (count != -1) {
+               enabledBeans.add(bean);
+               beforeGcCount += bean.getCollectionCount();
+            }
+         }
+
+         // Run the GC
+         this.runSystemGC();
+
+         // Make sure the GC actually happened
+         //   a) That at least two collections happened, indicating GC work.
+         //   b) That counter updates have not happened for a while, indicating GC work had ceased.
+
+         final int MAX_WAIT_MS = 10 * 1000;
+
+         if (enabledBeans.isEmpty()) {
+            log.warn("MXBeans can not report GC info. System.gc() invoked, but cannot check whether GC actually happened!");
+            return;
+         }
+
+         boolean gcHappened = false;
+         boolean stable = false;
+
+         long checkStart = System.nanoTime();
+         while (TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - checkStart) < MAX_WAIT_MS) {
+            long afterGcCount = 0;
+            for (GarbageCollectorMXBean bean : enabledBeans) {
+               afterGcCount += bean.getCollectionCount();
+            }
+
+            if (!gcHappened) {
+               if (afterGcCount - beforeGcCount >= 2) {
+                  gcHappened = true;
+               }
+            } else {
+               if (afterGcCount == beforeGcCount) {
+                  // Stable!
+                  stable = true;
+                  break;
+               }
+               beforeGcCount = afterGcCount;
+            }
+
+            try {
+               TimeUnit.MILLISECONDS.sleep(20);
+            } catch (InterruptedException e) {
+               Thread.currentThread().interrupt();
+            }
+         }
+
+         if (!stable) {
+            if (gcHappened) {
+               log.warn("System.gc() was invoked but unable to wait while GC stopped, is GC too asynchronous?");
+            } else {
+               log.warn("System.gc() was invoked but couldn't detect a GC occurring, is System.gc() disabled?");
+            }
+         }
+      }
+
+      log.debug("Overall GC run took {} ms", (System.currentTimeMillis() - start));
+   }
+
+   /**
+    * Executes the garbage collector (GC) and forces object finalization before each GC run.
+    * The method runs the garbage collector twice to ensure maximum garbage collection,
+    * and forces finalization of objects before each GC execution. It logs the time taken
+    * to complete the garbage collection process.
+    *
+    * @see System#runFinalization()
+    * @see System#gc()
+    */
+   private void runSystemGC() {
+      long actualGCRun = System.currentTimeMillis();
+
+      // Run the GC twice and force finalization before every GC run
+      System.runFinalization();
+      System.gc();
+      System.runFinalization();
+      System.gc();
+
+      log.debug("GC execution took {} ms", (System.currentTimeMillis() - actualGCRun));
```

Comment:
Why debug?

Reply:
Good catch, it was a leftover.. fixed now!